### DISCUSSION TITLE
ValueError: Detected non-contiguous batches.

### DISCUSSION THREAD

**Discussion by wangshuangyi** [brianhie/scanorama#131] [Q&A]
Hi @brianhie, many thanks for the great tool. However, I got a bit lost and hope you can help me out. I am trying to integrate some PBMC datasets from GEO and after preprocessing and adding batch information, an error is reported. I hope I can get your answer!

`Traceback (most recent call last):
  File "/home/data/vip08/imap/big_scrna/pan_read_csv/scanorma/scanorama_46Wpbmc.py", line 12, in <module>
    sce.pp.scanorama_integrate(data, key = 'x')
  File "/home/data/vip08/miniconda3/envs/imap/lib/python3.7/site-packages/scanpy/external/pp/_scanorama_integrate.py", line 110, in scanorama_integrate
    raise ValueError('Detected non-contiguous batches.')
ValueError: Detected non-contiguous batches.`

`data.obs['batch']`
`GSE138266/GSM4104134AGCTCCTGTGCTTCTC-1-12                batch6
GSE149689/ATGTCTTTCGATACTG-12-22                         batch7
GSE178318GCTGCAGGTAGTAGTA_COL16_CRC-84                  batch11
GSE96583/GSE96583_2_2/TTAGAATGCTCTTA-1-27                batch9
GSE178318CGTTCTGGTTAAGACA_COL16_CRC-84                  batch11
                                                         ...   
GSE157278/TTGAACGGTGAGTGAC-6-10                          batch4
GSE135779/GSE135779_sample_910/ATCATGGTCTACTTAC-1-42    batch10
GSE150861/CGAGAAGTCGAATGCT-2-0                           batch1
GSE135779/GSE135779_sample_904/ATTACTCGTAAGTGGC-1-36    batch10
GSE135779/GSE135779_sample_927/GAGGTGAGTGATGATA-1-59    batch10
`
`Name: batch, Length: 50449, dtype: category
Categories (11, object): ['batch1', 'batch2', 'batch3', 'batch4', ..., 'batch8', 'batch9', 'batch10', 'batch11']`

**Comment by brianhie** [comment]
Hi @wangshuangyi, this error happens if the batches are not contiguous, e.g., cells 1, 2, and 3 belong to batch A, B, and A, respectively. This could also happen if two of the batches actually have the same name. Could you check that cells belonging to the same batch are contiguous in the combined matrix?

**Reply by wangshuangyi** [reply]
Hi @brianhie, many thanks for your help! I checked and found that the cells belonging to the same batch in the combined matrix are not contiguous, the reason is that I took the subsampling step. I would like to ask you how can I make the batches contiguous?

**Reply by andrewdchen** [reply]
Sorry to revive an old thread, but I'm running into the same error. Also unclear what it means to make the batches contiguous? Is it a matter of rearranging the order of cells so that those that are in the same batch are next to each other?

**Reply by brianhie** [reply]
Hi @andrewdchen, yes this means keeping the cells "next to each other", e.g., [ "batch1", "batch1", "batch1", "batch2", "batch2", "batch2" ] instead of [ "batch1", "batch2", "batch1", "batch2", ... ].

You can reformat an anndata by doing something like the following pseudocode (I haven't tested this):
```python
adata_batch1 = adata[adata.obsm['X_batch'] == 'batch1']
adata_batch2 = adata[adata.obsm['X_batch'] == 'batch2']

adata_new = concat_function([ adata_batch1, adata_batch2 ])
```

**Reply by andrewdchen** [reply]
Thanks for the clarification @brianhie ! 
Since I had a lot of batches I just sorted, and re-indexed, seems to have the same effect.
```
idx = adata.obs.sort_values("batch").index
adata_new = adata[idx,]
```
